syntax = "proto3";

message VoteRequest {
  int32 candidate_id = 1;
  int32 term = 2;
  int32 log_length = 3;
  int32 last_term = 4;
}
message VoteResponse {
  int32 voter_id = 1;
  int32 term = 2;
  bool vote_granted = 3;
}

message LogRequest{
  message LogItem {
    string command = 1;
    string key = 2;
    string value = 3;
    int32 term = 4;
  }
  
  int32 leader_id = 1;
  int32 term = 2;
  int32 prefixLen = 3;
  int32 prefixTerm = 4;
  int32 leaderCommit = 5;
  repeated LogItem suffix = 6;
  int32 follower_id = 7;
}
message LogResponse{
  int32 node_id = 1;
  int32 term = 2;
  int32 ack = 3;
  bool success = 4;
}

//Client operations
message SetValRequest {
  string key = 1;
  string value = 2;
}

message SetValResponse {
  bool success = 1;
  int32 current_leader = 2;
}

message GetValRequest {
  string key = 1;
}

message GetValResponse {
  string value = 1;
  bool success = 2;
  int32 current_leader = 3;
}

service Node {
  rpc Log(LogRequest) returns (LogResponse);
  rpc Vote(VoteRequest) returns (VoteResponse);
  rpc SetVal(SetValRequest) returns (SetValResponse);
  rpc GetVal(GetValRequest) returns (GetValResponse);
}
